CMR-4103: Added functionality to respect all temporal ranges for coll…

sxu123 opened the pull request:
…ection search.

mschmele commented on indexer-app/src/cmr/indexer/data/concepts/collection.clj:
```diff
@@ -49,6 +50,9 @@
       :else
       (errors/internal-error! (str "Unknown spatial representation [" coord-sys "]")))))
 
+(defn- update-values [m f & args]
```
A docstring and more descriptive variable names would make this function a lot easier to read

indiejames replied:
A more descriptive function name would be helpful as well.

sxu123 replied:
Will add doctoring to describe it more. but I think update-values is all I need. because I just want to update all the values in a map.

sxu123 replied:
done.

sxu123 commented:
1. Confirmed with Erich that we currently don't have any collections that contain the periodic temporals.
   So, I only dealt with the RangeDateTime and SingleDateTime cases.
2. Added fake nested temporals for the granule searches because they share the same code as the collection searches in complex_to_simple_converters.
3. Added nested  limit-to-granules-temporals as a special case to respect all the collection temporal ranges when there are no granules for the collection and the limit_to_granules is set.

mschmele commented on indexer-app/src/cmr/indexer/data/concepts/collection.clj:
```diff
@@ -194,7 +207,15 @@
+        permitted-group-ids (get-coll-permitted-group-ids context provider-id collection)
+        {:keys [granule-start-date granule-end-date]} (cgac/get-coll-gran-aggregates context concept-id)
+        last-3-days (t/interval (t/minus (tk/now) (t/days 3)) (tk/now))
+        granule-end-date (if (and granule-end-date (t/within? last-3-days granule-end-date))
```
You could change this `if` to use `when-not`
https://clojuredocs.org/clojure.core/when-not

sxu123 replied:
I will take a look. Just cut and pasted from another section of the file.

indiejames commented on indexer-app/src/cmr/indexer/data/concepts/collection.clj:
```diff
@@ -183,6 +187,15 @@
+                              ;; converts each temporal-extent into a list of 
+                              ;; {:BeginningDateTime xxx :EndingDateTime xxx}
+                              (map spec-time/temporal-ranges)
+                              (flatten)
```
I think you can just call `mapcat` instead of `map` followed by `flatten`.

sxu123 replied:
yeah. thanks.

sxu123 replied:
done.

indiejames commented on indexer-app/src/cmr/indexer/data/concepts/collection.clj:
```diff
@@ -234,6 +255,14 @@
+            :limit-to-granules-temporals 
+              (if granule-start-date
+                [{:start-date (index-util/date->elastic granule-start-date)
+                  :end-date (index-util/date->elastic granule-end-date)}]
```
Does a `granule-end-date` of `nil` mean no end date? More importantly, does `date->elastic` handle nil correctly?

sxu123 replied:
I think so. I just moved this part of functionality from another section of the code which is working.

indiejames commented on umm-spec-lib/src/cmr/umm_spec/time.clj:
```diff
@@ -4,6 +4,21 @@
+        ;; otherwise, set :EndingDateTime to nil if it doesn't exist.
+        normalized-ranges (if (:EndsAtPresentFlag temporal) 
+                            (map #(assoc % :EndingDateTime nil) ranges)
+                            (map #(if (= nil (:EndingDateTime %)) (assoc % :EndingDateTime nil) %) ranges))
```
This line seems a little strange. Why is it necessary to set the value for the key `:EndingDateTime` to `nil` if it is already `nil` or not in the map? Also, you should use `nil?` instead of `(= nil` here.

sxu123 replied:
Since the result of this will need to map to the nested mapping which contains both the start-date and end-date, I just want to make sure both BeginningDateTime and EndingDateTime are present, even with nil value. I don't know what would happen to the nested mapping if an element is missing.

sxu123 replied:
done.

indiejames commented on indexer-app/src/cmr/indexer/data/concepts/collection.clj:
```diff
@@ -49,6 +50,9 @@
       (errors/internal-error! (str "Unknown spatial representation [" coord-sys "]")))))
 
+(defn- update-values [m f & args]
+  (reduce (fn [r [k v]] (assoc r k (apply f v args))) {} m))
```
Nit: using `reduce-kv` instead of `reduce` here will allow you to avoid destructuring the key/value pair in the anonymous function.

sxu123 replied:
Thanks. will look into that.

sxu123 replied:
done.

sxu123 commented:
This branch is messed up, replacing it with a new branch.
